src/mlagility/api/execute_ort.py

```diff
@@ -7,89 +7,126 @@
+        "--name",
+        docker_name,
+        docker_image,
+    ]
+    run_subprocess(cmd)
+
+
+def execute_benchmark(onnx_file, docker_name, num_iterations):
+    """Execute the benchmark script in a docker container and retrieve the output."""
+    cmd = (
+        "docker exec "
+        f"{docker_name} "
+        "/usr/bin/python3 /app/run_ort_model.py "
+        f"--onnx-file {onnx_file} "
+        f"--iterations {num_iterations}"
+    )
 
-    for _ in range(num_iterations):
-        start = timer()
+    with subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True
+    ) as proc:
+        stdout, stderr = proc.communicate()
+        if proc.returncode != 0 or not stdout:
+            raise ValueError(f"Execution of command {cmd} failed with stderr: {stderr}")
         try:
-            onnx_session.run([output_name], input_feed)
-        except Exception as e:  # pylint: disable=broad-except
-            exception = e
-        end = timer()
-        iteration_latency = end - start
-        per_iteration_latency.append(iteration_latency)
-
-    return per_iteration_latency, exception
-
-
-def _dummy_inputs(sess_input) -> dict:
-    # Generate dummy inputs of the expected shape and type for the input model
-    input_stats = []
-    for _idx, input_ in enumerate(range(len(sess_input))):
-        input_name = sess_input[input_].name
-        input_shape = sess_input[input_].shape
-
-        # TODO: Use onnx update_inputs_outputs_dims to automatically freeze models
-        for dim in input_shape:
-            if isinstance(dim, str) is True or math.isnan(dim) is True:
-                raise AssertionError(
-                    "Error: Model has dynamic inputs. Freeze the graph and try again"
-                )
-
-        input_type = sess_input[input_].type
-        input_stats.append([input_name, input_shape, input_type])
-
-    input_feed = {}
-    for stat in input_stats:
-        dtype_str = re.search(r"\((.*)\)", stat[2])
-        assert dtype_str is not None
-        datatype = dtype_ort2str(dtype_str.group(1))
-        input_feed[stat[0]] = np.random.rand(*stat[1]).astype(datatype)
-    return input_feed
-
-
-def dtype_ort2str(dtype_str: str):
-    if dtype_str == "float16":
-        datatype = "float16"
-    if dtype_str == "float":
-        datatype = "float32"
-    if dtype_str == "double":
-        datatype = "float64"
-    if dtype_str == "long":
-        datatype = "int64"
-    else:
-        datatype = dtype_str
-    return datatype
+            output_list = json.loads(stdout.decode("utf-8").strip())
+        except json.JSONDecodeError as e:
+            raise ValueError(
+                f"Failed to parse the output {stdout.decode('utf-8').strip()} as a list: {e}"
+            )
+        return output_list
+
+
+def stop_docker_container(docker_name):
+    """Stop and remove the docker container with the given name."""
+    cmd = ["docker", "stop", docker_name]
+    run_subprocess(cmd)
```

Comment:
(nonblocking) question: What happens if the docker container freezes (hangs forever)? Do we have a timeout function?

Reply:
A `timeout` option is available. But given we can benchmark any model, setting a fixed timeout might be dangerous. For e.g. if timeout=300s, it could also end up killing a model that takes 300s to run 100 iterations on.

Reply:
Good to know that a timeout function is available. I recommend setting it to 2h now for safety (same as in `slurm`). We can set it more appropriately later.

This will avoid the problem described in https://github.com/groq/mlagility/issues/171 (but I would still not close the issue, as we might want to address it in a nicer way).